include/pybind11/detail/class.h

```diff
@@ -388,15 +388,20 @@ inline void clear_patients(PyObject *self) {
     auto *instance = reinterpret_cast<detail::instance *>(self);
     auto &internals = get_internals();
     auto pos = internals.patients.find(self);
+
     assert(pos != internals.patients.end());
-    // Clearing the patients can cause more Python code to run, which
-    // can invalidate the iterator. Extract the vector of patients
-    // from the unordered_map first.
-    auto patients = std::move(pos->second);
-    internals.patients.erase(pos);
-    instance->has_patients = false;
-    for (PyObject *&patient : patients) {
-        Py_CLEAR(patient);
+    if (pos != internals.patients.end()) {
+        // Clearing the patients can cause more Python code to run, which
+        // can invalidate the iterator. Extract the vector of patients
+        // from the unordered_map first.
+        auto patients = std::move(pos->second);
+        internals.patients.erase(pos);
+        instance->has_patients = false;
+        for (PyObject *&patient : patients) {
+            Py_CLEAR(patient);
+        }
+    } else {
```

Comment:
Could you please simply replace the `assert`?

You can leave the rest of the code as is. Note that `pybind11_fail` is marked with `[[noreturn]]`.

Reply:
(This is called the “guard pattern”)

Reply:
What I had in mind was simply this net diff (nothing else):

```
--- a/include/pybind11/detail/class.h
+++ b/include/pybind11/detail/class.h
@@ -388,7 +388,9 @@ inline void clear_patients(PyObject *self) {
     auto *instance = reinterpret_cast<detail::instance *>(self);
     auto &internals = get_internals();
     auto pos = internals.patients.find(self);
-    assert(pos != internals.patients.end());
+    if (pos == internals.patients.end()) {
+        pybind11_fail("FATAL: Internal consistency check failed: Invalid clear_patients() call.");
+    }
     // Clearing the patients can cause more Python code to run, which
     // can invalidate the iterator. Extract the vector of patients
     // from the unordered_map first.
```

I would find it acceptable to move the existing code into an `else` block, but because `pybind11_fail` is marked with `[[noreturn]]` that's really not necessary and only makes the code less readable (IMO).

Placing `pybind11_fail` in the `else` block is unfortunate style IMO.

Reply:
@rwgk I changed it to this and will test it on trunk gcc tomorrow to make sure it doesn't get the same warning as before.